### PR TITLE
allow participants to access dataset through events

### PR DESCRIPTION
**Pull request by superryeti**
Context: This PR implements the last part of https://github.com/MIT-LCP/physionet-build/issues/1839
Here we will allow users who have been accepted to events to view the project/datasets.
The change should also allow HDN to access the project in the research environment.

Quick notes
1. So far the change doesnot make use of `access_type` from `EventDataset` model. It currently will let users registered to events access the dataset through Research environment(as the hdn-reserch-environment uses `PublishedProject.objects.accessible_by(user)`) , and view the files on the project page.

Screenshots to explain userstory
1. before user joins waitlist

image

2. After user joins waitlist
image

3. After user is approved to join the event
image

4. project page if user visits the project page through the correct link example(http://localhost:8000/content/demoeicu/2.0.0/?event=qoL47ZfEZjKw)
image

PS: if the user who has access to the dataset through event, visits the project through normal link http://localhost:8000/content/demoeicu/2.0.0 , they can still access the dataset and see files, but wont see the message.

I didnot restrict access to user when the link doesnot have `event=slug` because it might cause issues in cases where same user has access to the project through event and in normal way(credentialing and training).

---

**Explanation of the changes/Workflow, users and projects involved(or how everything is supposed to happen /happens)**

1. Event is first created by an Event Host
2. Event host will request the Platform Admin to add one or more dataset(s) to the Event
3. Platform Admin will then link one or more published project/dataset to the Event.
4. At this point the Event Host can access all the projects/dataset linked to the Event(until the Event ends or until the Admin unlinks the dataset/project from the Event)
5. The Event Host will share the event link to the students/participants, and the participants can request to join the event through the event link(the participants donot have access to project at this point).
6. Now the Event Host has the authority to grant access to dataset to the students by approving their request to join the event.
7. As soon as the Event Host approves students to join the event, student will have access the the project/dataset

**In Summary**

The Platform Admin has the highest Control, they have the ability to grant or revoke access to the project through the admin console.
After the Platform Admin, the Event host has the ability to let participants access the dataset by approving them to join the event.
Any dataset that is linked by the Platform Admin to an Event will be instantly accessible to the Host and Participants of that Event
The access expires after the event ends or if the Admin removes the project/dataset from Event.

**Quick note:**
There is also supposed to be further control on how the data is supposed to be accessed (currently)
1. Through Research Environment
2. Or through Google BigQuery
The current PR only works for the Research Environment at the moment.

### CODE REVIEW

**Comment by bemoody** [comment]
My initial reaction: this is exactly what I did not want to see at the *last minute before an event.*  This is a major change and not something that can be done in a hurry.

We need to spec out exactly what this means - **who** is supposed to be allowed to authorize **whom** to access **what**.

**Comment by superryeti** [comment]
> We need to spec out exactly what this means - **who** is supposed to be allowed to authorize **whom** to access **what**.

Thanks @bemoody . Let me edit on the  PR description. I will ping you once i update the description

**Comment by superryeti** [comment]
> My initial reaction: this is exactly what I did not want to see at the _last minute before an event._ This is a major change and not something that can be done in a hurry.
> 
> We need to spec out exactly what this means - **who** is supposed to be allowed to authorize **whom** to access **what**.

@bemoody  i updated the description. Sorry if i misunderstood  or didnot answer what you want meant, if that is the case, could you please let me know.

**Comment by superryeti** [comment]
Update.
As per feedback from @alistairewj and @tompollard 

I now refactored the code to include simple change which is to update `accessible_by` method in PublishedProjectManager.

This method is used by the hdn-research-environment.


This should help and make things work for the datathon.

For the further changes, i will followup with @alistairewj and @tompollard

File: physionet-django/events/views.py [hunk 1]
```diff
@@ -210,11 +210,16 @@ def event_detail(request, event_slug):
 
     # Handle Event Registration  End #
```
Comment (tompollard) [inline physionet-django/events/views.py:211]: This comment isn't needed (not sure if any of the `Handle Event Datastore` comments are needed tbh)

File: physionet-django/project/managers/publishedproject.py [hunk 2]
```diff
@@ -48,4 +51,18 @@ def accessible_by(self, user):
                 contributor_review_with_access | credentialed_with_dua_signed
             )
 
+        # Event related logic
+        # get all projects that are accessible by user through Events(Used by `hdn-research-environment` package)
+        # get all active events that user is a participant/host of
```
Comment (tompollard) [inline physionet-django/project/managers/publishedproject.py:56]: I'm not convinced that these comments are useful (if anything, for me they are distracting). 

Perhaps having one comment at the top explaining the overall goal of the chunk would be worthwhile. The rest we can work out from the code.

File: physionet-django/events/models.py [hunk 3]
```diff
@@ -189,6 +189,27 @@ def is_accessible(self):
             return False
         return True
 
+    def has_access(self, user):
+        """
+        checks if the user has access to the dataset for the Event
```
Comment (tompollard) [inline physionet-django/events/models.py:194]: This docstring isn't particularly clear.

**Review by tompollard — APPROVED** [review]
Thanks Amit, this is an improvement on the original PR and I think it's safe to merge. There's some cleaning to do (and a bigger discussion about how to refactor the code used to control access to content) but we can deal with this later.

File: physionet-django/project/managers/publishedproject.py (hunk 2 above)
Comment (superryeti) [reply physionet-django/project/managers/publishedproject.py:56]: I agree about the comments. They were perhaps my way of making sure i dont miss anything(i normally write a ton of comments before code and work on the code). I will refactor all of the unnecessary docstring, comments in another PR.

File: physionet-django/events/views.py (hunk 1 above)
Comment (superryeti) [reply physionet-django/events/views.py:211]: > Thanks Amit, this is an improvement on the original PR and I think it's safe to merge. There's some cleaning to do (and a bigger discussion about how to refactor the code used to control access to content) but we can deal with this later.


Noted. I will keep a list of needed changes and open a small cleanup PR after the datathon.

Also will catchup with you and Alistair to plan the refactoring of access and remaining/other feature of events.

**Comment by bemoody** [comment]
Would somebody be so kind as to post your private discussions regarding this pull request *to this pull request*?

**Comment by tompollard** [comment]
@bemoody there have been no discussions since we met at around 5.30pm (so I assume that when @amitupreti said "As per feedback from @alistairewj and @tompollard", he really meant feedback from you, me, and Alistair, via Alistair).

I looked over this PR and it seemed fairly innocuous, particularly now that the all access-related changes are isolated to PublishedProjectManager.  I'm sorry if you don't feel the same? Ideally would have discussed with you first...just timing has made things a bit tricky.

**Comment by superryeti** [comment]
> @bemoody there have been no discussions since we met at around 5.30pm (so I assume that when @amitupreti said "As per feedback from @alistairewj and @tompollard", he really meant feedback from you, me, and Alistair, via Alistair).
> 
> I looked over this PR and it seemed fairly innocuous, particularly now that the all access-related changes are isolated to PublishedProjectManager. I'm sorry if you don't feel the same? Ideally would have discussed with you first...just timing has made things a bit tricky.

Sorry yes, That is True, Alistair had told me he reviewed the PR with others. So i assumed it was you and him

**Comment by superryeti** [comment]
Here is the feedback i got from @alistairewj (Passed on by @tompollard and @bemoody )

(1) don't touch has_access. only provide access using accessible_by
(2) the redirect link currently goes to the project page, but instead, we can just link to the research environment page

The idea here was to only give access to the research environments to HDN, which uses accessible_by and plan about the rest after datathon.
